### PR TITLE
Clean up DataFrame.setitem behavior for duplicate columns

### PR DESCRIPTION
**Pull request by phofl**
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

cc  @jbrockmendel This cleans the edgy behavior up for duplicate columns.

Previous test was wrong.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3211,6 +3211,14 @@ def _setitem_slice(self, key: slice, value):
         self._check_setitem_copy()
         self.iloc[key] = value
 
+    def _check_key_length(self, key, value: DataFrame):
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3214]: maybe put in core.indexers and pass self.columns?

can this be reused anywhere else? (IIRC you recently implemented another get_indexer_non_unique check in a similar context)

File: pandas/tests/frame/indexing/test_setitem.py [hunk 2]
```diff
@@ -378,11 +378,19 @@ def test_setitem_df_wrong_column_number(self, cols):
     def test_setitem_listlike_indexer_duplicate_columns(self):
         # GH#38604
         df = DataFrame([[1, 2, 3]], columns=["a", "b", "b"])
-        rhs = DataFrame([[10, 11, 12]], columns=["d", "e", "c"])
+        rhs = DataFrame([[10, 11, 12]], columns=["a", "b", "b"])
```
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:381]: this seems more obvious.  should we also be testing the ["d", "e", "c"] case, or is that not well-defined?

File: pandas/core/frame.py (hunk 1 above)
Comment (jreback) [reply pandas/core/frame.py:3214]: +1 for this. ideally we move more and more indexing code out of here

File: pandas/tests/frame/indexing/test_setitem.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:381]: This aligns, so sets the values to nan. There is a discussion to deprecate this behavior (#39373). So did not want to add tests here

File: pandas/core/frame.py (hunk 1 above)
Comment (phofl) [reply pandas/core/frame.py]: Moved.

Unfortunately they only have the get_indexer_for call in common. The equality check is quite different

File: pandas/core/indexers.py [hunk 3]
```diff
@@ -376,6 +376,25 @@ def unpack_1tuple(tup):
     return tup
 
 
+def check_key_length(columns, key, value):
+    """Checks if a key used as indexer has the same length as the columns it is
```
Comment (jbrockmendel) [inline pandas/core/indexers.py:380]: newline after """

File: pandas/core/indexers.py [hunk 4]
```diff
@@ -376,6 +376,25 @@ def unpack_1tuple(tup):
...
+    associated with.
+
+    Parameters
+    ----------
+    columns: The columns of the DataFrame to index.
```
Comment (jbrockmendel) [inline pandas/core/indexers.py:385]: `columns : Index`

File: pandas/core/indexers.py [hunk 5]
```diff
@@ -376,6 +376,25 @@ def unpack_1tuple(tup):
...
+    key: A list-like of keys to index with.
+    value: The value to set for the keys.
+
+    """
+    if columns.is_unique:
+        if len(value.columns) != len(key):
+            raise ValueError("Columns must be same length as key")
+    else:
+        if len(columns.get_indexer_non_unique(key)[0]) != len(value.columns):
```
Comment (jbrockmendel) [inline pandas/core/indexers.py:394]: do we necessarily have `all(col in columns for col in key)`?

**Comment by jbrockmendel** [comment]
If you're interested, there might be something worth salvaging from _setitem_array [here](https://github.com/jbrockmendel/pandas/blob/setitem-frame-no-defer/pandas/core/frame.py#L3244).  Avoids going through iloc so that we dont have to worry about inplace vs not-inplace there (xref #38896).

If/when I get #39302 in shape to be a real PR, something like that branch will have to be a part of it, but I'm optimistic you'll get around to it before I do.

File: pandas/core/indexers.py (hunk 3 above)
Comment (phofl) [reply pandas/core/indexers.py:380]: Thx

File: pandas/core/indexers.py (hunk 4 above)
Comment (phofl) [reply pandas/core/indexers.py:385]: changed

File: pandas/core/indexers.py (hunk 5 above)
Comment (phofl) [reply pandas/core/indexers.py:394]: No, but this would return -1 as indexer for this col. I should probably add a note about this :)

**Comment by phofl** [comment]
That looks good. I will try to incorporate this after this is merged.

File: pandas/core/indexers.py [hunk 6]
```diff
@@ -376,6 +376,27 @@ def unpack_1tuple(tup):
...
+    """
+    Checks if a key used as indexer has the same length as the columns it is
+    associated with.
+
+    Parameters
+    ----------
+    columns: Index The columns of the DataFrame to index.
+    key: A list-like of keys to index with.
+    value: The value to set for the keys.
+
```
Comment (jreback) [inline pandas/core/indexers.py:389]: can you add a Raises section

File: pandas/core/indexers.py [hunk 7]
```diff
@@ -376,6 +376,27 @@ def unpack_1tuple(tup):
     return tup
 
 
+def check_key_length(columns, key, value):
```
Comment (jreback) [inline pandas/core/indexers.py:379]: can you type as much as you can here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexers.py [hunk 8]
```diff
@@ -376,6 +376,27 @@ def unpack_1tuple(tup):
...
+    """
+    Checks if a key used as indexer has the same length as the columns it is
+    associated with.
+
+    Parameters
+    ----------
+    columns: Index The columns of the DataFrame to index.
```
Comment (jbrockmendel) [inline pandas/core/indexers.py:386]: space between "columns" and colon, same on the next two lines

Comment (phofl) [reply pandas/core/indexers.py]: Thx

File: pandas/core/indexers.py (hunk 6 above)
Comment (phofl) [reply pandas/core/indexers.py:389]: Done

File: pandas/core/indexers.py (hunk 7 above)
Comment (phofl) [reply pandas/core/indexers.py:379]: We would have to type wiht Index and DataFrame causing circular imports. Is there a way around this?

Comment (phofl) [reply pandas/core/indexers.py:379]: Sorry, should have said that I have already tried this. If using TYPE_CHECKING and typing the signature (not the return value) this raises
NameError: name 'DataFrame' is not defined

Comment (phofl) [reply pandas/core/indexers.py:379]: Did not know about the future import. Thx for the help

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl 

was this technically a bug fix? or just an incorrect test?

**Comment by phofl** [comment]
Bug fix yes, but is covered by the whatsnew from #39341, follow up from that one

**Pull request merged**